### ISSUE TITLE
Support independence assumptions about params

### ISSUE THREAD

**Issue by fritzo** [pyro-ppl/pyro#1213]
We'd like to support batched second-order optimizers, where the loss Hessian has user-declared block diagonal structure.

One natural way to declare the param independence assumption is to define `param`s inside `iarange` contexts. Whereas the old issue #238 suggests adding broadcasting behavior to param-inside-iarange, this new issue suggests adding a semantic assumption that the resulting ELBO optimization problem should factor into blocks. Note that this assumption is a no-op when using first-order optimizers, and only arises when using higher-order optimizers like `newton_step()`.

The motivating use case is nested inference with an EM inner loop inside an SVI outer loop. In this case we've often hand-implemented the EM inner loop to solve a batch of low-dimensional problems via `newton_step()`. A more Pyronic way would be to implement the M-step using Pyro's `TraceEnum_ELBO`, something like
```py
def model_for_m_step(data, discrete_probs_from_e_step):
    with pyro.iparam("independent_subproblems", theta_dim):
        theta = pyro.param("theta", ...)
        assert theta.shape[-1] == theta_dim
        z = pyro.sample("discrete_variable",
                        dist.Categorical(discrete_probs_from_e_step),
                        infer={"enumerate": "parallel"})
        pyro.sample("obs", dist.Something(theta), obs=data[z])
        
def guide_for_m_step(data, discrete_probs_from_e_step):
    pass

elbo = TraceEnum_ELBO(max_iarange_nesting=1)
for _ in range(em_steps):
    discrete_probs_from_e_step = ...
    loss = elbo.differentiable_loss(model_for_m_step,
                                    guide_for_m_step,
                                    data, discrete_probs_from_e_step)
    theta_old = pyro.param("theta")
    theta_new = theta_old + newton_step(loss, theta_old)
    pyro.get_parm_store().replace_param("theta", theta_old, theta_new)
```

**Comment by neerajprad** [comment]
@fritzo, @eb8680: I was trying to understand what changes would we need in SVI, but I'm likely missing a few things:

> this new issue suggests adding a semantic assumption that the resulting ELBO optimization problem should factor into blocks. 

If `pyro.param("theta", ...)` still returns the global `theta`, then whether the optimization problem factorizes into sub-components would be determined by what parts of `theta` the subsequent `pyro.sample(..)` statements using them index into. Correct me if I am mistaken. Since the model above isn't using `theta`, its not very clear to me.

I was also wondering about potential incompatibility with vectorized num particles. Since they wrap the model and guide (and hence all `param` statements) inside an `iarange`, we wouldn't want that to affect any of the independence assumptions. This case can be addressed specifically though, and addressing it in the general case might not be that important. 

Let me know your thoughts. We can discuss in person too, when we are all here.

**Comment by eb8680** [comment]
We should just discuss in person next week, but re: incompatibility, we can initially cordon this functionality off inside a `Messenger` so that it only applies when we turn it on.

**Comment by fritzo** [comment]
> potential incompatibility with vectorized num particles

This feature is separate from `iarange`. Let's call it `iparam` or something. The behavior should be to (1) declare independent subproblems and (2) possibly also broadcast the `pyro.param` statement (this could alternatively be handled by `iarange`, but as @neerajprad mentioned that may conflict with parallel particles). As @eb8680 suggested, `iparam` can be a `Messenger`.

**Comment by fritzo** [comment]
> Since the model above isn't using theta

I've updated the model to use `theta`, but yes let's discuss in person.

**Comment by neerajprad** [comment]
Thanks for explaining, @fritzo. That makes sense, I'll try to grok this some more. :) Lets discuss in person later.

**Comment by fritzo** [comment]
This should now be possible after the introduction of an `event_dim` arg to `pyro.param` in #1796 . The discussion in this issue might be useful if we return to the issue of second-order optimizers.
